### PR TITLE
Port global caching strategy from Polymer 1.x

### PR DESCRIPTION
**Pull request by azakus**

### CODE REVIEW

File: src/ShadyCSS.js [hunk 1]
```diff
@@ -204,12 +202,19 @@ export let ShadyCSS = {
   },
   _applyStyleProperties(host, styleInfo) {
     let is = host.getAttribute('is') || host.localName;
+    let cacheEntry = styleCache.fetch(is, styleInfo.styleProperties, styleInfo.ownStylePropertyNames);
+    let cachedScopeSelector = cacheEntry && cacheEntry.scopeSelector;
+    let cachedStyle = cacheEntry ? cacheEntry.stylesheet : null;
```
Comment (sorvell) [inline src/ShadyCSS.js]: `stylesheet` => `styleElement`

File: src/style-cache.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+Code distributed by Google as part of the polymer project is also
+subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+*/
+'use strict';
+
+export default class StyleCache {
+  constructor(max = 100) {
+    // map element name -> [{properties, stylesheet, scopeSelector}]
+    this.cache = {};
+    this.max = max;
```
Comment (sorvell) [inline src/style-cache.js]: Let's name this more explicitly since it's a maximum per type, perhaps `typeMax` ?

File: src/style-cache.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    for (let idx = 0; idx < ownPropertyNames.length; idx++) {
+      let pn = ownPropertyNames[idx];
+      if (cacheEntry.properties[pn] !== properties[pn]) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  store(tagname, properties, stylesheet, scopeSelector) {
```
Comment (sorvell) [inline src/style-cache.js]: `stylesheet` => `styleElement`

File: src/ShadyCSS.js (hunk 1 above)
Comment (azakus) [inline src/ShadyCSS.js]: Done.

File: src/style-cache.js (hunk 2 above)
Comment (azakus) [inline src/style-cache.js]: Done.

File: src/style-cache.js (hunk 3 above)
Comment (azakus) [inline src/style-cache.js]: Done.

**Comment by azakus** [comment]
RSLGTM from @sorvell

**Pull request merged**